MAX7219 In Dot Matrix Configuration

<!-- READ THIS FIRST:
 - This is for feature requests only, if you want to have a certain new sensor/module supported, please use the "new integration" template.
 - Please be as descriptive as possible, especially use-cases that can otherwise not be solved boost the problem's priority.
 
   DO NOT DELETE ANY TEXT from this template! Otherwise the issue may be closed without a comment.
-->

**Is your feature request related to a problem/use-case? Please describe.**
<!-- add max7219 dot matrix support. -->
add max7219 dot matrix support.

**Describe the solution you'd like:**
<!-- home assistan publishing "events,massages.. on max7219 dot matrix . -->
home assistan publishing "events,massages.. on max7219 dot matrix .

**Additional context:**
<!-- sending text to display on dot matrix. -->
sending text to display on dot matrix.

@rspaargaren 
Great - I have cloned the branch and linked the max7219digit folder into the components folder of my installation.

There's a few warnings concerning max7219digit.cpp during compiling and all pixels in the display is on after flashing.

I use a Wemos D1 mini (probably a clone)

Virtualenv:
  Python 3.7.7
  ESPHome 1.14.3

```
INFO Reading configuration /var/db/esphome/demo1.yaml...
INFO Generating C++ source...
INFO Compiling app...
INFO Running:  platformio run -d /var/db/esphome/disco
Processing disco (board: d1_mini; framework: arduino; platform: espressif8266@2.2.3)
--------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
Dependency Graph
|-- <ESPAsyncTCP-esphome> 1.2.2
|   |-- <ESP8266WiFi> 1.0
|-- <ESP8266WiFi> 1.0
|-- <ESP8266mDNS> 1.2
|   |-- <ESP8266WiFi> 1.0
|-- <ESPAsyncWebServer-esphome> 1.2.6
|   |-- <ESPAsyncTCP-esphome> 1.2.2
|   |   |-- <ESP8266WiFi> 1.0
|   |-- <Hash> 1.0
|   |-- <ESP8266WiFi> 1.0
|-- <DNSServer> 1.1.1
|   |-- <ESP8266WiFi> 1.0
|-- <SPI> 1.0
Compiling .pioenvs/disco/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/disco/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/disco/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/disco/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/disco/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/disco/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/disco/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/disco/src/esphome/components/api/user_services.cpp.o
Compiling .pioenvs/disco/src/esphome/components/api/util.cpp.o
Compiling .pioenvs/disco/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling .pioenvs/disco/src/esphome/components/display/display_buffer.cpp.o
Compiling .pioenvs/disco/src/esphome/components/gpio/switch/gpio_switch.cpp.o
Compiling .pioenvs/disco/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/disco/src/esphome/components/max7219digit/max7219digit.cpp.o
Compiling .pioenvs/disco/src/esphome/components/ota/ota_component.cpp.o
Compiling .pioenvs/disco/src/esphome/components/spi/spi.cpp.o
Compiling .pioenvs/disco/src/esphome/components/switch/automation.cpp.o
src/esphome/components/max7219digit/max7219digit.cpp: In member function 'uint8_t esphome::max7219digit::MAX7219Component::print(uint8_t, const char*)':
src/esphome/components/max7219digit/max7219digit.cpp:119:1: warning: no return statement in function returning non-void [-Wreturn-type]
 }  // end of sendString
 ^
Compiling .pioenvs/disco/src/esphome/components/switch/switch.cpp.o
Compiling .pioenvs/disco/src/esphome/components/web_server_base/web_server_base.cpp.o
Compiling .pioenvs/disco/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/disco/src/esphome/components/wifi/wifi_component_esp32.cpp.o
Compiling .pioenvs/disco/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling .pioenvs/disco/src/esphome/core/application.cpp.o
Compiling .pioenvs/disco/src/esphome/core/component.cpp.o
Compiling .pioenvs/disco/src/esphome/core/controller.cpp.o
Compiling .pioenvs/disco/src/esphome/core/esphal.cpp.o
Compiling .pioenvs/disco/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/disco/src/esphome/core/log.cpp.o
Compiling .pioenvs/disco/src/esphome/core/preferences.cpp.o
Compiling .pioenvs/disco/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/disco/src/esphome/core/util.cpp.o
Compiling .pioenvs/disco/src/main.cpp.o
Generating LD script .pioenvs/disco/ld/local.eagle.app.v6.common.ld
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/BearSSLHelpers.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/CertStoreBearSSL.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/ESP8266WiFi.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/ESP8266WiFiAP.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/ESP8266WiFiGeneric.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/ESP8266WiFiSTA-WPS.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/ESP8266WiFiScan.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/WiFiClient.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/WiFiClientSecureAxTLS.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/WiFiClientSecureBearSSL.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/WiFiServer.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/WiFiServerSecureAxTLS.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/WiFiServerSecureBearSSL.cpp.o
Compiling .pioenvs/disco/lib2a8/ESP8266WiFi/WiFiUdp.cpp.o
Compiling .pioenvs/disco/lib36a/ESPAsyncTCP-esphome_ID6757/AsyncPrinter.cpp.o
Compiling .pioenvs/disco/lib36a/ESPAsyncTCP-esphome_ID6757/ESPAsyncTCP.cpp.o
Compiling .pioenvs/disco/lib36a/ESPAsyncTCP-esphome_ID6757/ESPAsyncTCPbuffer.cpp.o
Archiving .pioenvs/disco/lib2a8/libESP8266WiFi.a
Indexing .pioenvs/disco/lib2a8/libESP8266WiFi.a
Compiling .pioenvs/disco/lib36a/ESPAsyncTCP-esphome_ID6757/SyncClient.cpp.o
Compiling .pioenvs/disco/lib36a/ESPAsyncTCP-esphome_ID6757/tcp_axtls.c.o
Compiling .pioenvs/disco/lib0e5/ESP8266mDNS/ESP8266mDNS.cpp.o
Compiling .pioenvs/disco/lib0e5/ESP8266mDNS/ESP8266mDNS_Legacy.cpp.o
Compiling .pioenvs/disco/lib0e5/ESP8266mDNS/LEAmDNS.cpp.o
Archiving .pioenvs/disco/lib36a/libESPAsyncTCP-esphome_ID6757.a
Indexing .pioenvs/disco/lib36a/libESPAsyncTCP-esphome_ID6757.a
Compiling .pioenvs/disco/lib0e5/ESP8266mDNS/LEAmDNS_Control.cpp.o
Compiling .pioenvs/disco/lib0e5/ESP8266mDNS/LEAmDNS_Helpers.cpp.o
Compiling .pioenvs/disco/lib0e5/ESP8266mDNS/LEAmDNS_Structs.cpp.o
Compiling .pioenvs/disco/lib0e5/ESP8266mDNS/LEAmDNS_Transfer.cpp.o
Compiling .pioenvs/disco/libafc/Hash/Hash.cpp.o
Compiling .pioenvs/disco/libafc/Hash/sha1/sha1.c.o
Compiling .pioenvs/disco/lib39f/ESPAsyncWebServer-esphome_ID6758/AsyncEventSource.cpp.o
Archiving .pioenvs/disco/libafc/libHash.a
Indexing .pioenvs/disco/libafc/libHash.a
Compiling .pioenvs/disco/lib39f/ESPAsyncWebServer-esphome_ID6758/AsyncWebSocket.cpp.o
Compiling .pioenvs/disco/lib39f/ESPAsyncWebServer-esphome_ID6758/SPIFFSEditor.cpp.o
Archiving .pioenvs/disco/lib0e5/libESP8266mDNS.a
Indexing .pioenvs/disco/lib0e5/libESP8266mDNS.a
Compiling .pioenvs/disco/lib39f/ESPAsyncWebServer-esphome_ID6758/WebAuthentication.cpp.o
Compiling .pioenvs/disco/lib39f/ESPAsyncWebServer-esphome_ID6758/WebHandlers.cpp.o
Compiling .pioenvs/disco/lib39f/ESPAsyncWebServer-esphome_ID6758/WebRequest.cpp.o
Compiling .pioenvs/disco/lib39f/ESPAsyncWebServer-esphome_ID6758/WebResponses.cpp.o
Compiling .pioenvs/disco/lib39f/ESPAsyncWebServer-esphome_ID6758/WebServer.cpp.o
Compiling .pioenvs/disco/lib089/DNSServer/DNSServer.cpp.o
Archiving .pioenvs/disco/lib089/libDNSServer.a
Indexing .pioenvs/disco/lib089/libDNSServer.a
Compiling .pioenvs/disco/lib109/SPI/SPI.cpp.o
Archiving .pioenvs/disco/libFrameworkArduinoVariant.a
Indexing .pioenvs/disco/libFrameworkArduinoVariant.a
Compiling .pioenvs/disco/FrameworkArduino/Esp-frag.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/Esp-version.cpp.o
Archiving .pioenvs/disco/lib39f/libESPAsyncWebServer-esphome_ID6758.a
Indexing .pioenvs/disco/lib39f/libESPAsyncWebServer-esphome_ID6758.a
Compiling .pioenvs/disco/FrameworkArduino/Esp.cpp.o
Archiving .pioenvs/disco/lib109/libSPI.a
Indexing .pioenvs/disco/lib109/libSPI.a
Compiling .pioenvs/disco/FrameworkArduino/FS.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/FunctionalInterrupt.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/HardwareSerial.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/IPAddress.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/MD5Builder.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/Print.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/Schedule.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/ScheduledFunctions.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/StackThunk.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/Stream.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/StreamString.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/Tone.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/Updater.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/WMath.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/WString.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/abi.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/base64.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/cbuf.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/cont.S.o
Compiling .pioenvs/disco/FrameworkArduino/cont_util.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_app_entry_noextra4k.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_eboot_command.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_flash_utils.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_i2s.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_main.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_noniso.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_phy.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_postmortem.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_si2c.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_sigma_delta.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_timer.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_waveform.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_wiring.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_wiring_analog.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_wiring_digital.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_wiring_pulse.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_wiring_pwm.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/core_esp8266_wiring_shift.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/debug.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/gdb_hooks.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/heap.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/libb64/cdecode.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/libb64/cencode.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/libc_replacements.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/sntp-lwip2.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/spiffs/spiffs_cache.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/spiffs/spiffs_check.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/spiffs/spiffs_gc.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/spiffs/spiffs_hydrogen.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/spiffs/spiffs_nucleus.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/spiffs_api.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/spiffs_hal.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/sqrt32.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/time.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/uart.cpp.o
Compiling .pioenvs/disco/FrameworkArduino/umm_malloc/umm_malloc.cpp.o
Archiving .pioenvs/disco/libFrameworkArduino.a
Indexing .pioenvs/disco/libFrameworkArduino.a
Linking .pioenvs/disco/firmware.elf
Retrieving maximum program size .pioenvs/disco/firmware.elf
Checking size .pioenvs/disco/firmware.elf
Building .pioenvs/disco/firmware.bin
DATA:    [====      ]  44.4% (used 36388 bytes from 81920 bytes)
PROGRAM: [====      ]  39.0% (used 407240 bytes from 1044464 bytes)
Creating BIN file ".pioenvs/disco/firmware.bin" using ".pioenvs/disco/firmware.elf"
========================= [SUCCESS] Took 33.70 seconds =========================
INFO Successfully compiled program.
INFO Resolving IP address of disco.local
INFO  -> 192.168.1.149
INFO Uploading /var/db/esphome/disco/.pioenvs/disco/firmware.bin (411392 bytes)
Uploading: [============================================================] 100% Done...

INFO Waiting for result...
INFO OTA successful
INFO Successfully uploaded program.
INFO Starting log output from disco.local using esphome API
```

> Blx. @cvanaalst I got it wrong, just noticed !! The description is correct, the example is wrong !!
> 
> `config/esphome/custom_components/MAX7219digit/`
> 
> I've edited the original too.

Thanks 

but i see more errors while compiling
please help
```INFO Reading configuration /config/esphome/max7219digit.yaml...
INFO Generating C++ source...
INFO Compiling app...
INFO Running:  platformio run -d /config/esphome/max7219digit
Processing max7219digit (board: nodemcuv2; framework: arduino; platform: espressif8266@2.2.3)
--------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
Dependency Graph
|-- <ESPAsyncTCP-esphome> 1.2.2
|   |-- <ESP8266WiFi> 1.0
|-- <ESP8266WiFi> 1.0
|-- <ESP8266mDNS> 1.2
|   |-- <ESP8266WiFi> 1.0
|-- <ESPAsyncWebServer-esphome> 1.2.6
|   |-- <ESPAsyncTCP-esphome> 1.2.2
|   |   |-- <ESP8266WiFi> 1.0
|   |-- <Hash> 1.0
|   |-- <ESP8266WiFi> 1.0
|   |-- <ArduinoJson-esphomelib> 5.13.3
|-- <ArduinoJson-esphomelib> 5.13.3
|-- <DNSServer> 1.1.1
|   |-- <ESP8266WiFi> 1.0
|-- <SPI> 1.0
Compiling /data/max7219digit/.pioenvs/max7219digit/src/main.cpp.o
Compiling /data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o
Generating LD script /data/max7219digit/.pioenvs/max7219digit/ld/local.eagle.app.v6.common.ld
Compiling /data/max7219digit/.pioenvs/max7219digit/lib4d9/ESP8266WiFi/BearSSLHelpers.cpp.o
Compiling /data/max7219digit/.pioenvs/max7219digit/lib4d9/ESP8266WiFi/CertStoreBearSSL.cpp.o
In file included from src/main.cpp:18:0:
src/max7219digit.h:19:7: error: redefinition of 'class esphome::max7219digit::MAX7219Component'
 class MAX7219Component : public PollingComponent,
       ^
In file included from src/esphome.h:17:0,
                 from src/main.cpp:3:
src/esphome/components/max7219digit/max7219digit.h:19:7: error: previous definition of 'class esphome::max7219digit::MAX7219Component'
 class MAX7219Component : public PollingComponent,
       ^
In file included from /root/.platformio/packages/framework-arduinoespressif8266/tools/sdk/libc/xtensa-lx106-elf/include/sys/ctype.h:7:0,
                 from /root/.platformio/packages/framework-arduinoespressif8266/tools/sdk/libc/xtensa-lx106-elf/include/ctype.h:5,
                 from /root/.platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/4.8.2/cctype:42,
                 from /root/.platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/4.8.2/bits/localefwd.h:42,
                 from /root/.platformio/packages/toolchain-xtensa/xtensa-lx106-elf/include/c++/4.8.2/string:43,
                 from src/esphome/core/component.h:3,
                 from src/esphome/components/api/api_connection.h:3,
                 from src/esphome.h:2,
                 from src/main.cpp:3:
src/max7219font.h:3:46: error: redefinition of 'const byte MAX7219_Dot_Matrix_font [256][8]'
 const byte MAX7219_Dot_Matrix_font [256] [8] PROGMEM = {
                                              ^
In file included from src/esphome.h:18:0,
                 from src/main.cpp:3:
src/esphome/components/max7219digit/max7219font.h:3:12: error: 'const byte MAX7219_Dot_Matrix_font [256][8]' previously defined here
 const byte MAX7219_Dot_Matrix_font [256] [8] PROGMEM = {
            ^
Compiling /data/max7219digit/.pioenvs/max7219digit/lib4d9/ESP8266WiFi/ESP8266WiFi.cpp.o
*** [/data/max7219digit/.pioenvs/max7219digit/src/main.cpp.o] Error 1
========================== [FAILED] Took 2.51 seconds ==========================```

I see the following error after removing includes & compiling

```Archiving /data/max7219digit/.pioenvs/max7219digit/libFrameworkArduino.a
Linking /data/max7219digit/.pioenvs/max7219digit/firmware.elf
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::get_setup_priority() const':
max7219digit.cpp:(.text._ZNK7esphome12max7219digit16MAX7219Component18get_setup_priorityEv+0x4): multiple definition of `esphome::max7219digit::MAX7219Component::get_setup_priority() const'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZNK7esphome12max7219digit16MAX7219Component18get_setup_priorityEv+0x4): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::dump_config()':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component11dump_configEv+0x64): multiple definition of `esphome::max7219digit::MAX7219Component::dump_config()'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component11dump_configEv+0x64): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::update()':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component6updateEv+0x10): multiple definition of `esphome::max7219digit::MAX7219Component::update()'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component6updateEv+0x10): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::set_intensity(unsigned char)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component13set_intensityEh+0x0): multiple definition of `esphome::max7219digit::MAX7219Component::set_intensity(unsigned char)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component13set_intensityEh+0x0): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::set_num_chips(unsigned char)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component13set_num_chipsEh+0x0): multiple definition of `esphome::max7219digit::MAX7219Component::set_num_chips(unsigned char)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component13set_num_chipsEh+0x0): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::send_byte_(unsigned char, unsigned char)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component10send_byte_Ehh+0x8): multiple definition of `esphome::max7219digit::MAX7219Component::send_byte_(unsigned char, unsigned char)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component10send_byte_Ehh+0x8): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::display()':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component7displayEv+0x18): multiple definition of `esphome::max7219digit::MAX7219Component::display()'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component7displayEv+0x18): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::send_to_all_(unsigned char, unsigned char)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component12send_to_all_Ehh+0xc): multiple definition of `esphome::max7219digit::MAX7219Component::send_to_all_(unsigned char, unsigned char)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component12send_to_all_Ehh+0xc): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::setup()':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component5setupEv+0x30): multiple definition of `esphome::max7219digit::MAX7219Component::setup()'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component5setupEv+0x30): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::send64pixels(unsigned char, unsigned char const*)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component12send64pixelsEhPKh+0x14): multiple definition of `esphome::max7219digit::MAX7219Component::send64pixels(unsigned char, unsigned char const*)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component12send64pixelsEhPKh+0x14): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::sendSmooth(char const*, int)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component10sendSmoothEPKci+0x1c): multiple definition of `esphome::max7219digit::MAX7219Component::sendSmooth(char const*, int)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component10sendSmoothEPKci+0x1c): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::sendChar(unsigned char, unsigned char)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component8sendCharEhh+0xc): multiple definition of `esphome::max7219digit::MAX7219Component::sendChar(unsigned char, unsigned char)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component8sendCharEhh+0xc): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::print(unsigned char, char const*)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component5printEhPKc+0x8): multiple definition of `esphome::max7219digit::MAX7219Component::print(unsigned char, char const*)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component5printEhPKc+0x8): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::print(char const*)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component5printEPKc+0x4): multiple definition of `esphome::max7219digit::MAX7219Component::print(char const*)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component5printEPKc+0x4): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::printf(char const*, ...)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component6printfEPKcz+0x8): multiple definition of `esphome::max7219digit::MAX7219Component::printf(char const*, ...)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component6printfEPKcz+0x8): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::printf(unsigned char, char const*, ...)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component6printfEhPKcz+0x8): multiple definition of `esphome::max7219digit::MAX7219Component::printf(unsigned char, char const*, ...)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component6printfEhPKcz+0x8): first defined here
/data/max7219digit/.pioenvs/max7219digit/src/max7219digit.cpp.o: In function `esphome::max7219digit::MAX7219Component::set_writer(std::function<void (esphome::max7219digit::MAX7219Component&)>&&)':
max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component10set_writerEOSt8functionIFvRS1_EE+0x10): multiple definition of `esphome::max7219digit::MAX7219Component::set_writer(std::function<void (esphome::max7219digit::MAX7219Component&)>&&)'
/data/max7219digit/.pioenvs/max7219digit/src/esphome/components/max7219digit/max7219digit.cpp.o:max7219digit.cpp:(.text._ZN7esphome12max7219digit16MAX7219Component10set_writerEOSt8functionIFvRS1_EE+0x10): first defined here
collect2: error: ld returned 1 exit status
*** [/data/max7219digit/.pioenvs/max7219digit/firmware.elf] Error 1
========================= [FAILED] Took 14.28 seconds =========================```

Here's a log for your entertainment.

```
INFO Reading configuration /config/esphome/clock.yaml...
INFO Detected timezone 'GMT' with UTC offset 0 and daylight savings time from 28 March 01:00:00 to 31 October 02:00:00
INFO Generating C++ source...
INFO Compiling app...
INFO Running:  platformio run -d /config/esphome/clock
Processing clock (board: nodemcu-32s; framework: arduino; platform: espressif32@1.11.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
Dependency Graph
|-- <AsyncTCP-esphome> 1.1.1
|-- <ESPmDNS> 1.0
|   |-- <WiFi> 1.0
|-- <FS> 1.0
|-- <ESPAsyncWebServer-esphome> 1.2.6
|   |-- <AsyncTCP-esphome> 1.1.1
|   |-- <FS> 1.0
|   |-- <WiFi> 1.0
|-- <DNSServer> 1.1.0
|   |-- <WiFi> 1.0
|-- <Update> 1.0
|-- <WiFi> 1.0
|-- <SPI> 1.0
Compiling /data/clock/.pioenvs/clock/src/esphome/components/api/api_connection.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/api/api_pb2_service.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/api/api_server.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/api/list_entities.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/api/proto.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/api/subscribe_state.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/api/user_services.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/api/util.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/display/display_buffer.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/homeassistant/time/homeassistant_time.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/logger/logger.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/max7219digit/max7219digit.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/ota/ota_component.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/restart/restart_switch.cpp.o
Compiling /data/clock/.pioenvs/clock/src/esphome/components/spi/spi.cpp.o
src/esphome/components/max7219digit/max7219digit.cpp: In member function 'void esphome::max7219digit::MAX7219Component::scroll_left(uint8_t)':
src/esphome/components/max7219digit/max7219digit.cpp:147:3: error: 'uint8' was not declared in this scope
   uint8 NumSteps = stepsize + this -> stepsleft;
   ^
src/esphome/components/max7219digit/max7219digit.cpp:151:9: error: expected ';' before 'n'
   uint8 n = this->max_x+3;
         ^
In file included from src/esphome/components/max7219digit/max7219digit.cpp:2:0:
src/esphome/components/max7219digit/max7219digit.cpp:152:24: error: 'n' was not declared in this scope
   ESP_LOGD(TAG,"n: %i",n);
                        ^
src/esphome/core/log.h:88:89: note: in definition of macro 'esph_log_d'
   esp_log_printf_(ESPHOME_LOG_LEVEL_DEBUG, tag, __LINE__, ESPHOME_LOG_FORMAT(format), ##__VA_ARGS__)
                                                                                         ^
src/esphome/components/max7219digit/max7219digit.cpp:152:3: note: in expansion of macro 'ESP_LOGD'
   ESP_LOGD(TAG,"n: %i",n);
   ^
src/esphome/components/max7219digit/max7219digit.cpp:153:7: error: 'NumSteps' was not declared in this scope
   if (NumSteps >= this->max_x+3) 
       ^
src/esphome/components/max7219digit/max7219digit.cpp:155:21: error: 'NumSteps' was not declared in this scope
   this->stepsleft = NumSteps;
                     ^
src/esphome/components/max7219digit/max7219digit.cpp:157:13: error: expected ';' before 'j'
   for(uint8 j=1;j<NumSteps+1;j++){
             ^
src/esphome/components/max7219digit/max7219digit.cpp:157:17: error: 'j' was not declared in this scope
   for(uint8 j=1;j<NumSteps+1;j++){
                 ^
src/esphome/components/max7219digit/max7219digit.cpp:159:15: error: expected ';' before 'i'
     for(uint8 i=0;i<n-1;i++)
               ^
src/esphome/components/max7219digit/max7219digit.cpp:159:19: error: 'i' was not declared in this scope
     for(uint8 i=0;i<n-1;i++)
                   ^
*** [/data/clock/.pioenvs/clock/src/esphome/components/max7219digit/max7219digit.cpp.o] Error 1
========================= [FAILED] Took 114.05 seconds =========================
```

I have copied the latest version 
(https://github.com/rspaargaren/esphome/tree/MAX7219NEW/esphome/components/max7219digit) 
... and get a link error ...

HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash Dependency Graph |-- 
<ESPAsyncTCP-esphome> 1.2.2 | |-- <ESP8266WiFi> 1.0 |-- <ESP8266WiFi> 
1.0 |-- <ESP8266mDNS> 1.2 | |-- <ESP8266WiFi> 1.0 |-- <SPI> 1.0 
Compiling 
/data/max7219/.pioenvs/max7219/src/esphome/components/max7219digit/max7219digit.cpp.o 
Compiling /data/max7219/.pioenvs/max7219/src/main.cpp.o Linking 
/data/max7219/.pioenvs/max7219/firmware.elf 
/data/max7219/.pioenvs/max7219/src/main.cpp.o:(.text.setup+0x23c): 
undefined reference to 
`esphome::max7219digit::MAX7219Component::set_writer(std::function<void 
(esphome::max7219digit::MAX7219Component&)>&&)' 
/data/max7219/.pioenvs/max7219/src/main.cpp.o: In function `setup': 
main.cpp:(.text.setup+0x1b96): undefined reference to 
`esphome::max7219digit::MAX7219Component::set_writer(std::function<void 
(esphome::max7219digit::MAX7219Component&)>&&)' collect2: error: ld 
returned 1 exit status *** [/data/max7219/.pioenvs/max7219/firmware.elf] 
Error 1 ========================== [FAILED] Took 2.74 seconds 
==========================


chris


Op 10/05/2020 om 16:13 schreef rspaargaren:
>
> Hi in the latest code I have updated the limit to 100.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/esphome/feature-requests/issues/14#issuecomment-626334446>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AJTNOWIWNMMG754PHSJYNK3RQ2Y7LANCNFSM4GT34RKA>.
>
-- 
Chris van Aalst
cva@skynet.be

> I have copied the latest version (https://github.com/rspaargaren/esphome/tree/MAX7219NEW/esphome/components/max7219digit) ... and get a link error ... HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash Dependency Graph |-- <ESPAsyncTCP-esphome> 1.2.2 | |-- <ESP8266WiFi> 1.0 |-- <ESP8266WiFi> 1.0 |-- <ESP8266mDNS> 1.2 | |-- <ESP8266WiFi> 1.0 |-- <SPI> 1.0 Compiling /data/max7219/.pioenvs/max7219/src/esphome/components/max7219digit/max7219digit.cpp.o Compiling /data/max7219/.pioenvs/max7219/src/main.cpp.o Linking /data/max7219/.pioenvs/max7219/firmware.elf /data/max7219/.pioenvs/max7219/src/main.cpp.o:(.text.setup+0x23c): undefined reference to `esphome::max7219digit::MAX7219Component::set_writer(std::function<void (esphome::max7219digit::MAX7219Component&)>&&)' /data/max7219/.pioenvs/max7219/src/main.cpp.o: In function `setup': main.cpp:(.text.setup+0x1b96): undefined reference to `esphome::max7219digit::MAX7219Component::set_writer(std::function<void (esphome::max7219digit::MAX7219Component&)>&&)' collect2: error: ld returned 1 exit status *** [/data/max7219/.pioenvs/max7219/firmware.elf] Error 1 ========================== [FAILED] Took 2.74 seconds ========================== chris Op 10/05/2020 om 16:13 schreef rspaargaren:
> Hi in the latest code I have updated the limit to 100. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#14 (comment)](https://github.com/esphome/feature-requests/issues/14#issuecomment-626334446)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJTNOWIWNMMG754PHSJYNK3RQ2Y7LANCNFSM4GT34RKA>.
> -- Chris van Aalst cva@skynet.be

Ok I will do some more debugging. I had to polish the code to fullfill the requirements for the PULL REQUEST within ESP.

> I have copied the latest version (https://github.com/rspaargaren/esphome/tree/MAX7219NEW/esphome/components/max7219digit) ... and get a link error ... HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash Dependency Graph |-- <ESPAsyncTCP-esphome> 1.2.2 | |-- <ESP8266WiFi> 1.0 |-- <ESP8266WiFi> 1.0 |-- <ESP8266mDNS> 1.2 | |-- <ESP8266WiFi> 1.0 |-- <SPI> 1.0 Compiling /data/max7219/.pioenvs/max7219/src/esphome/components/max7219digit/max7219digit.cpp.o Compiling /data/max7219/.pioenvs/max7219/src/main.cpp.o Linking /data/max7219/.pioenvs/max7219/firmware.elf /data/max7219/.pioenvs/max7219/src/main.cpp.o:(.text.setup+0x23c): undefined reference to `esphome::max7219digit::MAX7219Component::set_writer(std::function<void (esphome::max7219digit::MAX7219Component&)>&&)' /data/max7219/.pioenvs/max7219/src/main.cpp.o: In function `setup': main.cpp:(.text.setup+0x1b96): undefined reference to `esphome::max7219digit::MAX7219Component::set_writer(std::function<void (esphome::max7219digit::MAX7219Component&)>&&)' collect2: error: ld returned 1 exit status *** [/data/max7219/.pioenvs/max7219/firmware.elf] Error 1 ========================== [FAILED] Took 2.74 seconds ========================== chris Op 10/05/2020 om 16:13 schreef rspaargaren:
> Hi in the latest code I have updated the limit to 100. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#14 (comment)](https://github.com/esphome/feature-requests/issues/14#issuecomment-626334446)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJTNOWIWNMMG754PHSJYNK3RQ2Y7LANCNFSM4GT34RKA>.
> -- Chris van Aalst cva@skynet.be

Hi, I did some changes to the code. Please check it should compile

rspaargaren,


it compiles and code executes correctly !

Op 10/05/2020 om 20:10 schreef rspaargaren:
>
>     I have copied the latest version
>     (https://github.com/rspaargaren/esphome/tree/MAX7219NEW/esphome/components/max7219digit)
>     ... and get a link error ... HARDWARE: ESP8266 80MHz, 80KB RAM,
>     4MB Flash Dependency Graph |-- 1.2.2 | |-- 1.0 |-- 1.0 |-- 1.2 |
>     |-- 1.0 |-- 1.0 Compiling
>     /data/max7219/.pioenvs/max7219/src/esphome/components/max7219digit/max7219digit.cpp.o
>     Compiling /data/max7219/.pioenvs/max7219/src/main.cpp.o Linking
>     /data/max7219/.pioenvs/max7219/firmware.elf
>     /data/max7219/.pioenvs/max7219/src/main.cpp.o:(.text.setup+0x23c):
>     undefined reference to
>     |esphome::max7219digit::MAX7219Component::set_writer(std::function<void
>     (esphome::max7219digit::MAX7219Component&)>&&)'
>     /data/max7219/.pioenvs/max7219/src/main.cpp.o: In function
>     |setup': main.cpp:(.text.setup+0x1b96): undefined reference to
>     `esphome::max7219digit::MAX7219Component::set_writer(std::function<void
>     (esphome::max7219digit::MAX7219Component&)>&&)' collect2: error:
>     ld returned 1 exit status ***
>     [/data/max7219/.pioenvs/max7219/firmware.elf] Error 1
>     ========================== [FAILED] Took 2.74 seconds
>     ========================== chris Op 10/05/2020 om 16:13 schreef
>     rspaargaren:
>     Hi in the latest code I have updated the limit to 100. — You are
>     receiving this because you were mentioned. Reply to this email
>     directly, view it on GitHub <#14 (comment)
>     <https://github.com/esphome/feature-requests/issues/14#issuecomment-626334446>>,
>     or unsubscribe
>     https://github.com/notifications/unsubscribe-auth/AJTNOWIWNMMG754PHSJYNK3RQ2Y7LANCNFSM4GT34RKA.
>     -- Chris van Aalst cva@skynet.be <mailto:cva@skynet.be>
>
> Hi, I did some changes to the code. Please check it should compile
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/esphome/feature-requests/issues/14#issuecomment-626366824>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AJTNOWPAOEDZM5GHPHXVOR3RQ3UYFANCNFSM4GT34RKA>.
>
-- 
Chris van Aalst
cva@skynet.be

rspaargaren,

I created my own arduino programs & modified existing programs (like 
MARQUEE) to display info on the MAX7219 display (esp8266); I am looking 
in converting some of these to esphome.

But what I miss is two things:

- speed-control: the native esp8266-MAX7219 scroll is faster & smoother: 
it scrolls per one dot - I assume this is not possible with esphome ?

- either longer messages (now limit 27+4), or an event that indicates 
the end of the line being displayed; in this case I could display a new 
line ? Other option is calculating how long it takes to display/scroll 
one character and display/scroll the new line (= not ideal, I used this 
solution in one of my first esp8266/max7219 versions, tricky and you 
should add spaces to start and end)

Additional:

- does intensity work ?

- I have different MAX7219 boards; the difference is how LED-matrix is 
implemented/oriented on board (of 4 matrices); some of the arduino 
MAX7219 libs can handle the different orientations: this might be 
interesting to add as an parameter.


He, I am very very happy with your work, it really works great and I 
will (and am) make use of it, the above are only my reflections on how 
to make it even better !!! (

again thx for super job,


Chris

Op 10/05/2020 om 20:10 schreef rspaargaren:
>
>     I have copied the latest version
>     (https://github.com/rspaargaren/esphome/tree/MAX7219NEW/esphome/components/max7219digit)
>     ... and get a link error ... HARDWARE: ESP8266 80MHz, 80KB RAM,
>     4MB Flash Dependency Graph |-- 1.2.2 | |-- 1.0 |-- 1.0 |-- 1.2 |
>     |-- 1.0 |-- 1.0 Compiling
>     /data/max7219/.pioenvs/max7219/src/esphome/components/max7219digit/max7219digit.cpp.o
>     Compiling /data/max7219/.pioenvs/max7219/src/main.cpp.o Linking
>     /data/max7219/.pioenvs/max7219/firmware.elf
>     /data/max7219/.pioenvs/max7219/src/main.cpp.o:(.text.setup+0x23c):
>     undefined reference to
>     |esphome::max7219digit::MAX7219Component::set_writer(std::function<void
>     (esphome::max7219digit::MAX7219Component&)>&&)'
>     /data/max7219/.pioenvs/max7219/src/main.cpp.o: In function
>     |setup': main.cpp:(.text.setup+0x1b96): undefined reference to
>     `esphome::max7219digit::MAX7219Component::set_writer(std::function<void
>     (esphome::max7219digit::MAX7219Component&)>&&)' collect2: error:
>     ld returned 1 exit status ***
>     [/data/max7219/.pioenvs/max7219/firmware.elf] Error 1
>     ========================== [FAILED] Took 2.74 seconds
>     ========================== chris Op 10/05/2020 om 16:13 schreef
>     rspaargaren:
>     Hi in the latest code I have updated the limit to 100. — You are
>     receiving this because you were mentioned. Reply to this email
>     directly, view it on GitHub <#14 (comment)
>     <https://github.com/esphome/feature-requests/issues/14#issuecomment-626334446>>,
>     or unsubscribe
>     https://github.com/notifications/unsubscribe-auth/AJTNOWIWNMMG754PHSJYNK3RQ2Y7LANCNFSM4GT34RKA.
>     -- Chris van Aalst cva@skynet.be <mailto:cva@skynet.be>
>
> Hi, I did some changes to the code. Please check it should compile
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/esphome/feature-requests/issues/14#issuecomment-626366824>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AJTNOWPAOEDZM5GHPHXVOR3RQ3UYFANCNFSM4GT34RKA>.
>
-- 
Chris van Aalst
cva@skynet.be

Thanks, sorry for late reply, was busy ...

I have tried compiling with the MAX7219NEW and I get this result.

```
Generating partitions /data/clock/.pioenvs/clock/partitions.bin
src/main.cpp: In lambda function:
src/main.cpp:712:12: error: 'class esphome::max7219digit::MAX7219Component' has no member named 'left_scroll'
         it.left_scroll(1);
            ^
Compiling /data/clock/.pioenvs/clock/lib219/AsyncTCP-esphome_ID6798/AsyncTCP.cpp.o
*** [/data/clock/.pioenvs/clock/src/main.cpp.o] Error 1
========================= [FAILED] Took 351.04 seconds =========================
```

> **Edit: I've changed ``` it.left_scroll ``` to ``` it.scroll_left ```  I've been trying to fix this for days ;)**
> 
> 
> 
> Thanks, sorry for late reply, was busy ...
> 
> 
> 
> I have tried compiling with the MAX7219NEW and I get this result.
> 
> 
> 
> ```
> 
> Generating partitions /data/clock/.pioenvs/clock/partitions.bin
> 
> src/main.cpp: In lambda function:
> 
> src/main.cpp:712:12: error: 'class esphome::max7219digit::MAX7219Component' has no member named 'left_scroll'
> 
>          it.left_scroll(1);
> 
>             ^
> 
> Compiling /data/clock/.pioenvs/clock/lib219/AsyncTCP-esphome_ID6798/AsyncTCP.cpp.o
> 
> *** [/data/clock/.pioenvs/clock/src/main.cpp.o] Error 1
> 
> ========================= [FAILED] Took 351.04 seconds =========================
> 
> ```

The devil is in the details... took me yesterday also about an hours to add one 0 to a code to fix it

INFO Reading configuration /config/esphome/led_matrix.yaml...
INFO Detected timezone 'KST' with UTC offset 9
INFO Generating C++ source...
INFO Compiling app...
INFO Running:  platformio run -d /config/esphome/led_matrix
Processing led_matrix (board: d1_mini; framework: arduino; platform: espressif8266@2.6.0)
--------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
PACKAGES: 
 - framework-arduinoespressif8266 3.20702.0 (2.7.2) 
 - tool-esptool 1.413.0 (4.13) 
 - tool-esptoolpy 1.20800.0 (2.8.0) 
 - toolchain-xtensa 2.40802.200502 (4.8.2)
Dependency Graph
|-- <ESPAsyncTCP-esphome> 1.2.2
|   |-- <ESP8266WiFi> 1.0
|-- <ESP8266WiFi> 1.0
|-- <ESP8266mDNS> 1.2
|   |-- <ESP8266WiFi> 1.0
|-- <SPI> 1.0
Compiling /data/led_matrix/.pioenvs/led_matrix/src/esphome/components/max7219digit/max7219digit.cpp.o
Compiling /data/led_matrix/.pioenvs/led_matrix/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling /data/led_matrix/.pioenvs/led_matrix/src/esphome/core/application.cpp.o
Compiling /data/led_matrix/.pioenvs/led_matrix/src/esphome/core/component.cpp.o
In file included from src/esphome/components/max7219digit/max7219digit.cpp:1:0:
src/esphome/components/max7219digit/max7219digit.h:43:51: error: 'Color' has not been declared
   void draw_absolute_pixel_internal(int x, int y, Color color) override;
                                                   ^
src/esphome/components/max7219digit/max7219digit.cpp:126:71: error: 'Color' has not been declared
 void HOT MAX7219Component::draw_absolute_pixel_internal(int x, int y, Color color) {
                                                                       ^
src/esphome/components/max7219digit/max7219digit.cpp: In member function 'virtual void esphome::max7219digit::MAX7219Component::draw_absolute_pixel_internal(int, int, int)':
src/esphome/components/max7219digit/max7219digit.cpp:137:13: error: request for member 'is_on' in 'color', which is of non-class type 'int'
   if (color.is_on()) {
             ^
*** [/data/led_matrix/.pioenvs/led_matrix/src/esphome/components/max7219digit/max7219digit.cpp.o] Error 1
========================== [FAILED] Took 5.56 seconds ==========================

--- a/esphome/components/max7219digit/display.py
+++ b/esphome/components/max7219digit/display.py
@@ -58,7 +58,7 @@ def to_code(config):
     cg.add(var.set_scroll_delay(config[CONF_SCROLL_DELAY]))
     cg.add(var.set_scroll(config[CONF_SCROLL_ENABLE]))
     cg.add(var.set_scroll_mode(config[CONF_SCROLL_MODE]))
-    cg.add(var.set_scroll(config[CONF_REVERSE_ENABLE]))
+    cg.add(var.set_reverse(config[CONF_REVERSE_ENABLE]))
 
     if CONF_LAMBDA in config:
         lambda_ = yield cg.process_lambda(config[CONF_LAMBDA], [(MAX7219ComponentRef, 'it')],

--- a/esphome/components/max7219digit/display.py
+++ b/esphome/components/max7219digit/display.py
@@ -11,6 +11,7 @@ CONF_SCROLL_DWELL = 'scroll_dwell'
 CONF_SCROLL_DELAY = 'scroll_delay'
 CONF_SCROLL_ENABLE = 'scroll_enable'
 CONF_SCROLL_MODE = 'scroll_mode'
+CONF_REVERSE_ENABLE = 'reverse_enable'
 
 SCROLL_MODES = {
     'CONTINUOUS': 0,
@@ -39,6 +40,7 @@ CONFIG_SCHEMA = display.BASIC_DISPLAY_SCHEMA.extend({
     cv.Optional(CONF_SCROLL_SPEED, default='250ms'): cv.positive_time_period_milliseconds,
     cv.Optional(CONF_SCROLL_DELAY, default='1000ms'): cv.positive_time_period_milliseconds,
     cv.Optional(CONF_SCROLL_DWELL, default='1000ms'): cv.positive_time_period_milliseconds,
+    cv.Optional(CONF_REVERSE_ENABLE, default=False): cv.boolean,
 }).extend(cv.polling_component_schema('500ms')).extend(spi.spi_device_schema(cs_pin_required=True))
 
 
@@ -56,6 +58,7 @@ def to_code(config):
     cg.add(var.set_scroll_delay(config[CONF_SCROLL_DELAY]))
     cg.add(var.set_scroll(config[CONF_SCROLL_ENABLE]))
     cg.add(var.set_scroll_mode(config[CONF_SCROLL_MODE]))
+    cg.add(var.set_scroll(config[CONF_REVERSE_ENABLE]))
 
     if CONF_LAMBDA in config:
         lambda_ = yield cg.process_lambda(config[CONF_LAMBDA], [(MAX7219ComponentRef, 'it')],

--- a/esphome/components/max7219digit/max7219digit.cpp
+++ b/esphome/components/max7219digit/max7219digit.cpp
@@ -109,7 +109,7 @@ void MAX7219Component::display() {
   for (uint8_t i = 0; i < this->num_chips_; i++) {
     for (uint8_t j = 0; j < 8; j++) {
       if (this->reverse_) {
-        pixels[j] = this->max_displaybuffer_[(this->num_chips_ - i -1) * 8 + j];
+        pixels[j] = this->max_displaybuffer_[(this->num_chips_ - i - 1) * 8 + j];
       } else {
         pixels[j] = this->max_displaybuffer_[i * 8 + j];
       }

--- a/esphome/components/max7219digit/max7219digit.cpp
+++ b/esphome/components/max7219digit/max7219digit.cpp
@@ -108,7 +108,11 @@ void MAX7219Component::display() {
   // Send the data to the chip
   for (uint8_t i = 0; i < this->num_chips_; i++) {
     for (uint8_t j = 0; j < 8; j++) {
-      pixels[j] = this->max_displaybuffer_[i * 8 + j];
+      if (this->reverse_) {
+        pixels[j] = this->max_displaybuffer_[(this->num_chips_ - i -1) * 8 + j];
+      } else {
+        pixels[j] = this->max_displaybuffer_[i * 8 + j];
+      }
     }
     this->send64pixels(i, pixels);
   }

--- a/esphome/components/max7219digit/max7219digit.h
+++ b/esphome/components/max7219digit/max7219digit.h
@@ -52,6 +52,7 @@ class MAX7219Component : public PollingComponent,
   void set_scroll_delay(uint16_t delay) { this->scroll_delay_ = delay; };
   void set_scroll(bool on_off) { this->scroll_ = on_off; };
   void set_scroll_mode(uint8_t mode) { this->scroll_mode_ = mode; };
+  void set_reverse(bool on_off) { this->reverse_ = on_off; };
 
   void send_char(byte chip, byte data);
   void send64pixels(byte chip, const byte pixels[8]);
@@ -87,6 +88,7 @@ class MAX7219Component : public PollingComponent,
   uint8_t intensity_;  /// Intensity of the display from 0 to 15 (most)
   uint8_t num_chips_;
   bool scroll_;
+  bool reverse_;
   bool update_{false};
   uint16_t scroll_speed_;
   uint16_t scroll_delay_;

